File: algorithms/python/235_lowest_common_ancestor_of_a_binary_search_tree/lowest_common_ancestor_of_a_binary_search_tree.py
"""
235. lowest common ancestor of a binary search tree/

二叉搜索树的最近公共祖先
"""


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def lowest_common_ancestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
    ancestor = root
    while True:
        if p.val < ancestor.val and q.val < ancestor.val:
            ancestor = ancestor.left
        elif p.val > ancestor.val and q.val > ancestor.val:
            ancestor = ancestor.right
        else:
            return ancestor

File: algorithms/235-lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.go
package problem235

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// lowestCommonAncestorV1 利用二叉搜索树的特性做两次遍历，
func lowestCommonAncestorV1(root, p, q *TreeNode) *TreeNode {
	pathP := getPath(root, p)
	pathQ := getPath(root, q)
	var ancestor *TreeNode
	for i := 0; i < len(pathP) && i < len(pathQ) && pathP[i] == pathQ[i]; i++ {
		ancestor = pathP[i]
	}
	return ancestor
}

func getPath(node, target *TreeNode) []*TreeNode {
	var path []*TreeNode
	for node != nil {
		path = append(path, node)
		if node.Val == target.Val {
			return path
		}
		if node.Val > target.Val {
			node = node.Left
		} else {
			node = node.Right
		}
	}
	return path
}

// lowestCommonAncestorV2 一次遍历
func lowestCommonAncestorV2(root, p, q *TreeNode) *TreeNode {
	ancestor := root
	for {
		if p.Val < ancestor.Val && q.Val < ancestor.Val {
			ancestor = ancestor.Left
		} else if p.Val > ancestor.Val && q.Val > ancestor.Val {
			ancestor = ancestor.Right
		} else {
			return ancestor
		}
	}
}
